### PR TITLE
Improve validate performance by removing unnecessary file IO

### LINKED ISSUE #358: Improve validate performance by removing unnecessary file IO
## Additional Details

When PDS4 IM explicitly includes file naming requirements, we should utilize that to improve performance. Finding collection and bundle files is a great example of this.

## Engineering Details

- [x] Use similar functionality or develop helper function similar to [Collection Integrity checker](https://github.com/NASA-PDS/validate/blob/18e5c566f7e6f26674c06d0813c88cb89579cc9f/src/main/java/gov/nasa/pds/tools/validate/rule/pds4/CollectionReferentialIntegrityRule.java#L79) and/or [Bundle Integrity checker](https://github.com/NASA-PDS/validate/blob/18e5c566f7e6f26674c06d0813c88cb89579cc9f/src/main/java/gov/nasa/pds/tools/validate/rule/pds4/BundleReferentialIntegrityRule.java#L308) to use filenames to check for bundles/collections instead of opening and reading the files

Per conversation via email:

If a user has a very large bundle, say more than a million labels and they just want to skip the product validation, we don’t validate the label against the schema and schematrons.  The check to see if it is a bundle or a collection is time consuming.
 
```
{pds-dev3.jpl.nasa.gov}/home/qchau/sandbox/validate 108 % view  src/main/java/gov/nasa/pds/tools/label/LabelValidator.java
 
   590       // Determine if schematron validation should be done or not.
    591       // Note: schematron validation can be time consuming.  Only the bundle or collection should be validated against schematron.
    592       // If the flag skipProductValidation is true, the labels belong to data files should not be done.
    593       Boolean validateAgainstSchematronFlag = this.determineSchematronValidationFlag(url);
    594       LOG.debug("parseAndValidate:url,skipProductValidation,validateAgainstSchematronFlag {},{},{}",url,skipProductValidation,validateAgainstSchematronFlag);
    595
    596 long t9_a = System.currentTimeMillis();
    597       for (Transformer schematron : cachedSchematron) {
    598         long singleSchematronStartTime = System.currentTimeMillis();
    599         if (!validateAgainstSchematronFlag) continue;  // Skip the validation if validateAgainstSchematronFlag is not true.
    600 long t5_a = System.currentTimeMillis();
    601         DOMResult result = new DOMResult();
    602         DOMSource domSource = new DOMSource(xml);
    603         LOG.debug("parseAndValidate:VALIDATING_SCHEMATRON_URL:START {} against schematron",url);
    604         domSource.setSystemId(url.toString());
    605 long t5_b = System.currentTimeMillis();
    606 t5_elapsed = t5_b - t5_a;
    607 LOG.debug("parseAndValidate:DOMSource:t5_elapsed {}",t5_elapsed);
    608         // Apply the rules specified in the schematron file
    609         schematron.transform(domSource, result);
    610         // Output is svrl:schematron-output document
    611         // Select out svrl:failed-assert nodes and put into problem container
    612         Document reportDoc = (Document) result.getNode();
    613         NodeList nodes = reportDoc.getElementsByTagNameNS(
    614             "http://purl.oclc.org/dsdl/svrl", "failed-assert");
    615         for (int i = 0; i < nodes.getLength(); i++) {
    616           Node node = nodes.item(i);
    617           // Add an error for each failed asssert
    618           handler.addProblem(processFailedAssert(url, node, xml));
    619         }
    620         long singleSchematronFinishTime = System.currentTimeMillis();
    621         long singleSchematronTimeElapsed = singleSchematronFinishTime - singleSchematronStartTime;
    622         LOG.debug("parseAndValidate:VALIDATING_SCHEMATRON_URL:ELAPSED {} {}",url,singleSchematronTimeElapsed);
    623       }
    624 long t9_b = System.currentTimeMillis();
```
 
So if the user wants does NOT want to skip product validation, everything is fine at line 360, it can return true.  If they want to skip product validation, it proceeds to 373 to open the file
to check to see if it is a bundle or a collection.  If it is then set to true to validate the label against the schema, if it is a regular label, set to false (don’t validate the schematron).  The check at line 373 is time consuming for
every label.   Perhaps we can change the code at line 372 where it checks for the name only after making everything lowercase and instead of using startsWith() it can use contains().
 
```
347   private Boolean determineSchematronValidationFlag(URL url) {
 348       // Function return true if the validation against the schematron should be done or not.
 349       // Note: schematron validation can be time consuming.  Only the bundle or collection should be validated against schematron.
 350       // If the flag skipProductValidation is true, the labels belong to data files should not be done.
 351       String nameOnly = Paths.get(url.getPath()).getFileName().toString();
 352 
 353       //LOG.debug("url.getPath() {}",url.getPath()); 
 354       //LOG.debug("nameOnly {}",nameOnly); 
 355       //LOG.debug("nameOnly.contains('bundle') {}",nameOnly.contains("bundle")); 
 356       //LOG.debug("nameOnly.contains('collection') {}",nameOnly.contains("collection")); 
 357 
 358       Boolean validateAgainstSchematronFlag;
 359       if (!this.skipProductValidation) {
 360           validateAgainstSchematronFlag = true;
 361       } else {
 362           // If skip product validation is true, perform schematron validation bundle/collection labels only.
 363           // Note: To inspect the file is time consuming
 364           // If the name of the file contains 'bundle' or 'collection' do the schematron validation.
 365           //if (TargetExaminer.isTargetBundleType(url) || TargetExaminer.isTargetCollectionType(url))
 366           // Comment out the file inspection in favor of checking the name only as it can be time consuming.
 367           // 6C.1.3 Reserved File Names
 368           //     The following file names are reserved for specific purposes within PDS archives and may not be used otherwise:
 369           //         bundle*.xml
 370           //         collection*.[xml,csv]
 371 
 372           //if ((nameOnly.startsWith("bundle") || nameOnly.startsWith("collection")) && nameOnly.endsWith(".xml")) {
 373           if (TargetExaminer.isTargetBundleType(url) || TargetExaminer.isTargetCollectionType(url)) {
 374               validateAgainstSchematronFlag = true;
 375           } else {
 376               // Do not validate labels belonging to data files.
 377               validateAgainstSchematronFlag = false;
 378           }
 379       }
 380       LOG.debug("url,validateAgainstSchematronFlag {},{}",url,validateAgainstSchematronFlag);
 381       return(validateAgainstSchematronFlag);
 382   }
```

### PR DESCRIPTION
**Pull request by qchaupds**
Improve validate performance by removing unnecessary file IO

1. Add BUNDLE_LABEL_PATTERN and COLLECTION_LABEL_PATTERN patterns in member variables for reuse :src/main/java/gov/nasa/pds/tools/label/LabelValidator.java
2. Replace time consuming check of reading from file to see if it is a bundle or a collection with name pattern check :src/main/java/gov/nasa/pds/tools/label/LabelValidator.java

Refs:

https://github.com/nasa-pds/validate/issues/358 Improve validate performance by removing unnecessary file IO

Closes https://github.com/nasa-pds/validate/issues/358 Improve validate performance by removing unnecessary file IO

The new code still works as before.  The new changes reflect better performance in checking if a label is a bundle or a collection.

Since issue #358 is an improvement on performance for bundle.  Any valid bundle can be used to test.

Tested in DEV:

Case 1:  Validate a valid bundle.

```
% validate -R pds4.bundle --skip-product-validation --skip-context-validation --skip-content-validation -r report_github358_bundle_valid.json -s json -t src/test/resources/github240/valid/bundle_kaguya_derived.xml >& t1
```
Please see comment of issue #358 for timing related findings.

### CODE REVIEW

File: src/main/java/gov/nasa/pds/tools/label/LabelValidator.java [hunk 1]
```diff
@@ -363,11 +366,20 @@ private Boolean determineSchematronValidationFlag(URL url) {
           //         bundle*.xml
           //         collection*.[xml,csv]
 
-          if (TargetExaminer.isTargetBundleType(url) || TargetExaminer.isTargetCollectionType(url)) {
+          // Remove time consuming check of reading the file in favor of checking the filename.
+          // The file name should be enough to know if it is a bundle or a collection.
+          validateAgainstSchematronFlag = false;
+          Matcher matcher = null;
```
[the remark this answers is not in the document]
Comment (qchaupds) [reply src/main/java/gov/nasa/pds/tools/label/LabelValidator.java:372]: ignore

[the remark this answers is not in the document]
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/tools/label/LabelValidator.java:372]: @sonatype-lift ignore

File: src/main/java/gov/nasa/pds/tools/label/LabelValidator.java [hunk 2]
```diff
@@ -363,11 +366,20 @@ private Boolean determineSchematronValidationFlag(URL url) {
...
+          matcher = BUNDLE_LABEL_PATTERN.matcher(FilenameUtils.getName(url.toString()));
```
Comment (jordanpadams) [inline src/main/java/gov/nasa/pds/tools/label/LabelValidator.java:373]: @qchaupds I don't think the string matching is a very expensive operation + it seems like we will continue to have nesting of checks here, but not a huge deal. we are a go for merge (after #396)

**Review by jordanpadams — APPROVED** [review]

File: src/main/java/gov/nasa/pds/tools/label/LabelValidator.java (hunk 2 above)
Comment (qchaupds) [reply src/main/java/gov/nasa/pds/tools/label/LabelValidator.java:373]: @jordanpadams The time consuming process before was actually opening and reading the label file twice.  Once was to check if it a bundle and if not open it again to check if it is a collection.  Now, we can just look at the file name.

**Pull request merged**